### PR TITLE
DOI and Snapshotting

### PR DESCRIPTION
**Pull request by david4096**
Putting together the DOI and snapshotting features.

### CODE REVIEW

**Comment by david4096** [comment]
Needs integration tests, having some trouble getting zenodo linking to work at the moment and so haven't tested it end to end.

image

image

image

**Comment by coverbeck** [comment]
Just going off the screenshots:

1. Your screenshots have 3 ways of referring to the version: `1.2.0`, `this version: 1.1.0`, and `this version 1.1.0`. I actually prefer a 4th way :), just simple`version x`, e.g., `are you sure you'd like to create a snapshot for version 1.2.0`, but whichever way you pick should be consistent.
1. I like the bold for the version.
1. You probably just haven't got to it yet, but the other thing we'd talked about was a DOI column.

BTW, because there are a fair amount of PRs floating around, I believe most people go off `Review Requests` in the GitHub UI (at least I typically do), so if you want more feedback (?) I would recommend assigning reviewer(s). Even though you mentioned this at standup and on Slack, most of us will just go off of what we see assigned in GitHub. I just happened to remember randomly. :)

**Comment by denis-yuen** [comment]
>  I believe most people go off `Review Requests` in the GitHub UI (at least I typically do), so if you want more feedback (?) I would recommend assigning reviewer(s). Even though you mentioned this at standup and on Slack, most of us will just go off of what we see assigned in GitHub. I just happened to remember randomly. :)

Worse(?) I usually just look at notifications so don't feel shy about unassigning/assigning people or re-requesting reviews.

**Comment by denis-yuen** [comment]
For the last screenshot, while DOIs cannot be undone, much of the data in the DOI can be edited since we're using the user's own credentials. Maybe the language should make that more clear and/or link to a page using an info button to readthedocs (can create a stub in advance and fill out later).

**Comment by david4096** [comment]
Opening this for a quick review/merge. Upcoming PR will add tests and better usability.

**Comment by david4096** [comment]
image

File: src/app/workflow/versions/versions.component.html [hunk 1]
```diff
@@ -150,13 +143,25 @@
...
+              *ngIf="!version.doiURL && !version.frozen"
+              data-cy="dockstore-snapshot-unlocked"
+              matTooltip="Version {{ version.name }} can be snapshotted."
+              >close</mat-icon
+            >
+            <a
+              href="https://doi.org/{{ version.doiURL }}"
+              *ngIf="version.doiURL"
+              matTooltip="Click to view this version's DOI entry information.."
+              ><img src="https://sandbox.zenodo.org/badge/DOI/{{ version.doiURL }}.svg" alt="DOI"
```
Comment (coverbeck) [inline src/app/workflow/versions/versions.component.html]: Should not be hard-coded to sandbox.

File: src/app/workflow/view/view.component.html [hunk 2]
```diff
@@ -40,3 +29,27 @@
...
 </ng-container>
+<button
+  type="button"
+  mat-button
+  color="accent"
+  class="btn-block"
+  *ngIf="!version.frozen && version.referenceType !== 'BRANCH'"
+  (click)="snapshotVersion()"
+  data-cy="dockstore-snapshot"
+  [matTooltip]="snapshotTooltip"
```
Comment (coverbeck) [inline src/app/workflow/view/view.component.html]: Should disable while snapshotting

File: src/app/workflow/view/view.component.html [hunk 3]
```diff
@@ -40,3 +29,27 @@
...
+>
+  Snapshot
+</button>
+<button
+  mat-button
+  id="DOIButton"
+  color="accent"
+  class="btn-block"
+  (click)="requestDOIForWorkflowVersion()"
+  [disabled]="isRefreshing$ | async"
```
Comment (coverbeck) [inline src/app/workflow/view/view.component.html]: Haven't gotten to TS code yet, but probably should be disabled on requestingDOI, not refreshing.

File: src/app/workflow/view/view.service.ts [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  ) {}
+
+  /**
+   * Updates the workflow version and alerts the Dockstore User with success
+   * or failure.
+   *
+   * @private
+   * @memberof ViewService
+   */
+  private updateWorkflowToSnapshot(workflow: Workflow, version: WorkflowVersion, cb: Function): void {
```
Comment (coverbeck) [inline src/app/workflow/view/view.service.ts]: The callback function is node.js-y. Should use Observable.

Comment (david4096) [reply src/app/workflow/view/view.service.ts]: I left in one way of doing this using observables for the snapshot. I'm not very happy with the refactor it made though. The callback method is still used for the DOI button for comparison (and to get advice).

File: src/app/workflow/view/view.service.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+      // Guarantee we don't snapshot a snapshot
+      return;
+    }
+    const dialogData: ConfirmationDialogData = {
+      message: `Snapshotting a version will make it so it can no longer be edited and cannot be undone. <p>Are you sure you would like to snapshot version <b>${version.name}</b>?`,
+      title: 'Snapshot',
+      confirmationButtonText: 'Snapshot Version',
+      cancelButtonText: 'Cancel'
+    };
+    return this.confirmationDialogService.openDialog(dialogData, '500px');
```
Comment (garyluu) [inline src/app/workflow/view/view.service.ts]: ctrl-f 500px

Use `bootstrap4mediumModalSize` for `500px`

File: src/app/workflow/view/view.component.html [hunk 6]
```diff
@@ -40,3 +29,28 @@
...
+>
+  Snapshot
+</button>
+<button
+  mat-button
+  id="DOIButton"
+  color="accent"
+  class="btn-block"
+  (click)="requestDOIForWorkflowVersion()"
+  [disabled]="doiRequested"
```
Comment (garyluu) [inline src/app/workflow/view/view.component.html]: I think simply disabling both of these with isRefreshing$ would've been sufficient and would've avoided all the other problems

:man_shrugging:

File: src/app/workflow/view/view.component.ts [hunk 7]
```diff
@@ -119,26 +125,56 @@ export class ViewWorkflowComponent extends View implements OnInit {
...
+    });
+  }
+
   /**
    * Opens a confirmation dialog that the Dockstore User can use to
    * confirm they want a snapshot.
    *
    * @memberof ViewWorkflowComponent
    */
   snapshotVersion(): void {
```
Comment (garyluu) [inline src/app/workflow/view/view.component.ts]: Move the bulk of this function into a service.  If you really want to keep this.snapshotRequested, just move that to the state

File: src/app/workflow/view/view.component.ts [hunk 8]
```diff
@@ -119,26 +125,56 @@ export class ViewWorkflowComponent extends View implements OnInit {
...
     );
   }
+
+  /**
+   * Handles the create DOI button being clicked
+   *
+   * @memberof ViewWorkflowComponent
+   */
+  requestDOIForWorkflowVersion() {
+    this.doiRequested = true;
```
Comment (garyluu) [inline src/app/workflow/view/view.component.ts]: if decided to not simply use isRefreshing$, move this into the state too

File: src/app/workflow/view/view.service.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    this.alertService.detailedSuccess();
+  }
+
+  /**
+   * Opens a confirmation dialog that the Dockstore User can use to
+   * confirm they want a snapshot.
+   *
+   * @memberof ViewService
+   */
+  snapshotVersion(workflow: Workflow, version: WorkflowVersion): Observable<any> {
```
Comment (garyluu) [inline src/app/workflow/view/view.service.ts]: observable boolean

File: src/app/workflow/view/view.component.html (hunk 6 above)
Comment (coverbeck) [reply src/app/workflow/view/view.component.html]: My bad then. When I pulled the code, I didn't even see a isRefreshing$ -- IDE as flagging it as a compile error.

**Pull request merged**